### DISCUSSION TITLE
Changes in new Marp Core v4 🚀

### DISCUSSION THREAD

**Discussion by yhatt** [marp-team/marp#533] [Announcements]
**[Marp Core](https://github.com/marp-team/marp-core)** has now been updated major version to **[v4.0.0](https://github.com/marp-team/marp-core/releases/tag/v4.0.0)**. This update brings a lot of improvements and updates.

### TL;DR: New Marp Core v4 is available in the new [Marp CLI v4](https://github.com/marp-team/marp-cli/releases/v4.0.0) and [Marp for VS Code v3](https://github.com/marp-team/marp-vscode/releases/v3.0.0).

# ⚡️ Breaking changes

## Node.js 18 and later is required

Marp Core v4 requires Node.js 18 or later. If you are using outdated Node.js 16 or earlier to run CLI or your app that is integrated with Marp, you need to upgrade Node.js.

## The slide element now has `display: block`

Until Marp Core v3, the slide element `<section>` was a [Flexbox](https://developer.mozilla.org/docs/Learn/CSS/CSS_layout/Flexbox) container `display: flex` in some themes and variants, `default`, `uncover`, and `gaia` with `lead` class. It was a required restriction to make content vertically centered.

However, it restricted some useful CSS features provided to [the basic box model](https://developer.mozilla.org/docs/Web/CSS/CSS_box_model/Introduction_to_the_CSS_box_model) too. Actually some Marp users were confused by the unavailability of that features. For example:

- [**Floating layout**](https://developer.mozilla.org/docs/Learn/CSS/CSS_layout/Floats): marp-team/marp#50
- [**CSS Multiple-column layout**](https://developer.mozilla.org/docs/Learn/CSS/CSS_layout/Multiple-column_Layout): https://github.com/marp-team/marp/discussions/192#discussioncomment-6023390
- [**Margin collapsing**](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_box_model/Mastering_margin_collapsing)
- etc.

[Baseline 2024](https://web.dev/baseline/2024) browsers have improved the block layout to be able to [center content vertically without using Flexbox](https://web.dev/blog/align-content-block). Therefore, the time has come to remove the confusing restriction, and change the slide container layout from `display: flex` to `display: block`.

By this change, Marp could be easier to use more CSS layout features. A following example is using the floating layout to the image:

```markdown
<!-- size: 4:3 -->

<img src="https://marp.app/favicon.png" alt="Marp" width="200" height="200" align="right" />

# **Marp**: Markdown Presentation Ecosystem

Marp helps your writing Markdown for creating the slide deck. Marp's simple and minimal design makes it easy to focus on your content, and the logical flow of the presentation.
```

<p align="center">
  <img src="https://github.com/user-attachments/assets/8b164124-5af5-478c-8bca-968d7c35a76a" alt="Floating layout" />
</p>

### ↩︎ Revert to Flexbox

This change **may break the existing slide with the container that is expected Flexbox**. In addition, **outdated browsers before [Baseline 2024](https://web.dev/baseline/2024) cannot render centered contents** in the block layout.

You can tweak the style in Markdown or the custom theme to get back the previous Flexbox container.

- For `default` and `uncover` theme:

  ```html
  <style>
    section {
      display: flex;
    }
  </style>
  ```

- For `gaia` theme's `lead` class:

  ```html
  <style>
    section.lead {
      display: flex;
    }
  </style>
  ```

When reverted to Flexbox, the slide container will make compatible with v3, and following "Safe centering" is also disabled.

## Safe centering

Introduces "**Safe centering**", to help creating the clear and well layouted slide deck for audiences.

Visual margins are important for the readability of a presentation. Every built-in theme has defined the padding, the space from the slide edge to the content area. That area would be better to avoid placing slide content, but sometimes overflowed from the safe area if put too much content on a single slide page.

Before v4, the vertical centering provided by built-in themes was always centered forcibly, even if the content might be overlapped with the padding area.

<p align="center">
  <img src="https://github.com/user-attachments/assets/70b8b379-926a-4f68-9cc8-f672c09b6916" alt="Unsafe centering" />
</p>

In Marp Core v4, content becomes centered vertically **within the safe area** by [`align-content: safe center`](https://developer.mozilla.org/docs/Web/CSS/align-content#safe). It will overflow from only the bottom of the safe area if a slide page has too much content.

<p align="center">
  <img src="https://github.com/user-attachments/assets/77053edf-c7b8-4cd5-b65f-61b191d6809b" alt="Safe centering" />
</p>

This change makes slide authors easier to pay attention to the padding area, and encourages creating visually better slides. It also helps to prevent from the bad habit for the readability, putting too much content on a single page.

### ↩︎ Revert to unsafe centering

If you want to revert to the previous centering, you can overload `align-content` property with `unsafe center` in your Markdown or the custom theme.

- For `default` and `uncover` theme:

  ```html
  <style>
    section {
      align-content: unsafe center;
    }
  </style>
  ```

- For `gaia` theme's `lead` class:

  ```html
  <style>
    section.lead {
      align-content: unsafe center;
    }
  </style>
  ```

# ✨ Updates

## Relaxed HTML usage

v4 has relaxed the default limitation of HTML usage in Markdown, and a lot of HTML elements and attributes are now allowed by default.

In Marp slides, HTML elements are commonly used within Markdown to fine-tune the layout. Many users had explicitly enabled an option to permit HTML elements in each Marp tool, but this has a big risk because it also enables insecure elements that must be used with caution.

In addition, we have discovered through issues and discussions that forgetting to permit HTML elements is a very common pitfall for Marp users. The previous Marp Core had been allowed only `<br>` tag by default, but we have concluded that it is desirable to permit enough HTML elements by default.

The full list of default allowed HTML elements and attributes is available in https://github.com/marp-team/marp-core/blob/main/src/html/allowlist.ts.

Many thanks to the feedback from the community in the discussion [marp-team/marp#501](https://github.com/orgs/marp-team/discussions/501).

> [!NOTE]
> The insecure HTML elements and attributes such as `<script>` element and `on****` attribute are not yet allowed. To allow all HTML usage including insecure, you should permit explicitly in the setting of Marp tools as before.

## Initial support for CSS nesting

Backed by the contribution to [Marpit framework](https://marpit.marp.app/) [v3.1.0](https://github.com/marp-team/marpit/releases/tag/v3.1.0), the initial support for [CSS nesting](https://developer.mozilla.org/docs/Web/CSS/CSS_nesting) has been added.

```html
<style>
  h1 {
    strong {
      color: red;

      &:hover {
        color: green;
      }
    }
  }
</style>

# Hello, **world**!
```

Marp will try to make flatten nested CSS rules before rendering, to get compatibility with CSS manipulations by the current Marpit. So the result of styling might become slightly different from native CSS nesting spec (e.g. the calculation of [specificity](https://developer.mozilla.org/docs/Web/CSS/Specificity)). We have planned to support CSS nesting in the next-gen framework correctly.

**Comment by alanlivio** [comment]
That is amazing @yhatt. 🥳🥳🥳
